idaes/dmf/getver.py

```diff
@@ -0,0 +1,218 @@
+import subprocess
+from types import ModuleType
+from typing import Union
+
+_log = logging.getLogger(__name__)
+
+
+# Exceptions
+
+
+class GetVersionError(Exception):
+    pass
+
+
+class ModuleImportError(GetVersionError):
+    pass
+
+
+class PipError(GetVersionError):
+    pass
+
+
+class GitHashError(GetVersionError):
+    pass
+
+
+# Classes and functions
+
+
+#: Return from `Versioned.get_info()`
+VersionInfo = namedtuple("VersionInfo", "version git_hash")
+
+
+class Versioned:
+    """Get version information for a module.
+
+    Sample usage::
+
+        from idaes.dmf.getver import Versioned
+        import some_module
+
+        vrs = Versioned(some_module)
+        print(vrs.version)
+        print(vrs.git_hash)
+
+    """
+
+    #: PIP command
+    PIP = "pip"
+
+    def __init__(self, obj: Union[ModuleType, str]):
+        """Constructor.
+
+        Args:
+            obj: Imported package or package name.
+        """
+        if hasattr(obj, "__package__"):
+            mod = obj
+        else:
+            try:
+                mod = importlib.import_module(obj)
+            except ModuleNotFoundError as err:
+                raise ModuleImportError(f"Could not import module '{obj}': {err}")
+        self._mod = mod
+        self._name = mod.__package__
+        try:
+            self._version = self._get_version()
+        except PipError:
+            raise GetVersionError(f"Could not get version for '{self._name}' from pip command '{self.PIP}'")
+        if self._version is None:
+            raise GetVersionError(f"Could not get version for '{self._name}'")
+        self._hash = None
+
+    def get_info(self) -> VersionInfo:
+        """Get hash and version without throwing any exceptions.
+        If there was an exception, the hash may be None.
+
+        Returns:
+            The Git hash (which may be None) and the version (a string)
+        """
+        hash = None
+        try:
+            hash = self.git_hash
+        except GetVersionError:
+            pass
+        return VersionInfo(git_hash=hash, version=self.version)
+
+    @property
+    def name(self) -> str:
+        """Get the name of the module.
+
+        Returns:
+            Module name.
+        """
+        return self._name
+
+    @property
+    def version(self) -> str:
+        """Get the version.
+
+        Return:
+            The version as a string, e.g. "2.1.0"
+        """
+        return self._version
+
+    @property
+    def git_hash(self) -> str:
+        """Get the git hash.
+
+        Returns:
+            Git hash as a string of hexadecimal digits, e.g. "a5327b889b4606792cd31194f939c0d358e68e43"
+
+        Raises:
+            GitHashError
+        """
+
+        if self._hash is None:
+            self._hash = self._get_git_hash()
+        return self._hash
+
+    # -- Protected methods -- #
+
+    def _get_version(self):
+        version = None
+        try:
+            version = pkg_resources.get_distribution(self._name).version
```

Comment:
Note that this will not work for IDAES: the package is `idaes-pse`, but `self._name == 'idaes'` (the namespace)